### PR TITLE
Migrate style exceptions to homebrew/core

### PR DESCRIPTION
**Pull request by Rylan12**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?
- [x] Have you successfully run `brew man` locally and committed any changes?

-----

This PR migrates the style exceptions to homebrew/core as part of the tap list migration. Corresponds with https://github.com/Homebrew/homebrew-core/pull/65927

There are two reasons why this migration is more tricky than the previous one. First, we cannot access the typical `Tap` methods in the `rubocop` code. That means that accessing the style exceptions is not as simple as running `tap.style_exceptions` like it would be elsewhere in the `brew` code. To work around this, I created a method called [`tap_style_exception?`](https://github.com/Rylan12/brew/blob/dfb59a1b5786c3bc87b5988d48c97bb72fd28c0a/Library/Homebrew/rubocops/extend/formula.rb#L476-L500) in `rubocops/extend/formula.rb`. This method will need to manually read the JSON files containing the rules, duplicating some code in [`tap.rb`](https://github.com/Rylan12/brew/blob/dfb59a1b5786c3bc87b5988d48c97bb72fd28c0a/Library/Homebrew/tap.rb#L573-L577). I've set it up, though, so the method will only need to read the files the first time it's called so this shouldn't be an efficiency issue. Let me know if there's a better solution to this.

Secondly, when testing this I've been running into some issues with `rubocop`'s caching. Modifying the tap style exception lists does cause `rubocop` to re-read, so it uses the old cache to check the formulae. This probably won't be an issue for CI, but for a user who's testing commands locally it will probably be annoying. They would run `brew style <formula>`, see an error, fix it by modifying a list, and rerun `brew style <formula>` only to see the same error that they just fixed. The way to get around this is to either modify the `rubocop` code to force a cache-reload or to manually run `rubocop --cache false <formula>`. This may not be a huge issue because, hopefully, these lists won't need to be modified very often, but it's definitely something to keep in mind.

Lastly, I want to comment on another change that I made to the [`uses_from_macos`](https://github.com/Rylan12/brew/blob/dfb59a1b5786c3bc87b5988d48c97bb72fd28c0a/Library/Homebrew/rubocops/uses_from_macos.rb) cop. I separated the two constants (`ALLOWED_USES_FROM_MACOS_DEPS` and `PROVIDED_BY_MACOS_FORMULAE`) into two separate lists and cops. The first cop, [`ProvidedByMacos`](https://github.com/Rylan12/brew/blob/dfb59a1b5786c3bc87b5988d48c97bb72fd28c0a/Library/Homebrew/rubocops/uses_from_macos.rb#L9-L19), checks that all formulae marked `keg_only :provided_by_macos` are added to the `provided_by_macos_formulae` list. This list is then used by the [`UsesFromMacos`](https://github.com/Rylan12/brew/blob/dfb59a1b5786c3bc87b5988d48c97bb72fd28c0a/Library/Homebrew/rubocops/uses_from_macos.rb#L21-L37) cop to check that `uses_from_macos` calls only reference formulae in the `provided_by_macos_formulae` and `non_keg_only_provided_by_macos_formulae` lists. This means that we don't have to worry about forgetting to update the `PROVIDED_BY_MACOS_FORMULAE` list anymore. Now, `brew style` will require updating those lists when adding a new `keg_only :provided_by_macos` formula.

### CODE REVIEW

File: Library/Homebrew/rubocops/extend/formula.rb [hunk 1]
```diff
@@ -472,6 +473,32 @@ def formula_tap
         match_obj[1]
       end
 
+      # Returns whether the given formula exists in the given style exception list.
+      # Defaults to the current formula being checked.
+      def tap_style_exception?(list, formula = nil)
+        if @tap_style_exceptions.blank? && formula_tap.present?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/extend/formula.rb:479]: Probably can't rely on ActiveSupport i.e. blank/present here 😭

**Review by MikeMcQuaid — APPROVED** [review]
Looks good so far! Another (simpler?) option may be to use the existing `#rubocop:disable` comments. I'm not sure if they are more or less desirable. I like the ability to be `CODEOWNERS` on some of these exceptions.

File: Library/Homebrew/rubocops/extend/formula.rb (hunk 1 above)
Comment (Rylan12) [reply Library/Homebrew/rubocops/extend/formula.rb:479]: Oh yeah, probably not. `.nil?` works, though, right?

**Comment by Rylan12** [comment]
Anyone in @Homebrew/core have an opinion?

The options seem to be:

1. Maintain these allowlists in the `style_exceptions/` directory (similar to the existing `audit_exceptions/` directory)
    - Pros: less clutter in formula files, it's somewhat protected against unwanted additions to allowlists (especially if we use `CODEOWNERS`), doesn't advertise that we allow for exceptions—we don't want to encourage users to just add an exception when things don't work
    - Cons: you need to edit multiple files to add an exception (although this could be a pro from the point of view of "we want to make it more difficult to add exceptions"), more tricky to implement (shouldn't really be a factor once it's done, though) plus the `brew style` caching issue
1. Choose to allow certain style failures by adding `#rubocop:disable` comments in formula files
    - Pros: no need for `style_exceptions/` directory, easier to maintain the lists because there aren't any, no need to worry about the `brew style` caching issue
    - Cons: looks less pretty (in my opinion), we don't want to encourage style exceptions so making it easier to ignore them might be a bad thing, can't protect the lists with a `CODEOWNERS` file

Another factor: we may still need to have these lists somewhere for the `UsesFromMacos` cop (unless there's another way to check or we ignore it). If we need lists for one, is it any harder to get lists working for the rest?

**Comment by Rylan12** [comment]
Okay, no responses so far (that's okay it's been a weekend) so here's my opinion:

I think we should continue with the exception lists method. I think there are clear benefits (e.g. having `CODEOWNERS`, helping to avoid adding unnecessary exceptions, and cleaner looks). The main downside is this rubocop cache issue. However, since we shouldn't be adding exceptions ver ofter, this shouldn't be a huge issue. I'm currently looking into adding a `--reset-cache` option to `brew style` which should also help.

**Comment by reitermarkus** [comment]
> Choose to allow certain style failures by adding `#rubocop:disable` comments in formula files

We specifically have a RuboCop enabled for official taps which prohibits the use of `rubocop:disable` comments. I think this should stay as-is.

**Comment by MikeMcQuaid** [comment]
> We specifically have a RuboCop enabled for official taps which prohibits the use of `rubocop:disable` comments. I think this should stay as-is.

I forgot about this. Given that I'd also agree with 1). 👍🏻

**Comment by Rylan12** [comment]
Okay, sounds good. Good to start moving forward by merging https://github.com/Homebrew/homebrew-core/pull/65927?

File: Library/Homebrew/rubocops/extend/formula.rb [hunk 2]
```diff
@@ -472,6 +473,32 @@ def formula_tap
...
 
+      # Returns whether the given formula exists in the given style exception list.
+      # Defaults to the current formula being checked.
+      def tap_style_exception?(list, formula = nil)
+        if @tap_style_exceptions.nil? && !formula_tap.nil?
+          @tap_style_exceptions = {}
+
+          style_exceptions_dir = "#{File.dirname(File.dirname(@file_path))}/style_exceptions/*.json"
+          Pathname.glob(style_exceptions_dir).each do |exception_file|
+            list_name = exception_file.basename.to_s.chomp(".json").to_sym
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/extend/formula.rb:484]: Might be nice to have a simple unit test for this.

File: Library/Homebrew/rubocops/uses_from_macos.rb [hunk 3]
```diff
@@ -6,82 +6,20 @@
 module RuboCop
   module Cop
     module FormulaAudit
+      # This cop audits formulae that are keg-only because they are provided by macos.
+      class ProvidedByMacos < FormulaCop
+        def audit_formula(_node, _class_node, _parent_class_node, body_node)
+          find_method_with_args(body_node, :keg_only, :provided_by_macos) do
+            unless tap_style_exception? :provided_by_macos_formulae
+              problem "Formulae that are `keg_only :provided_by_macos` should be added to "\
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/uses_from_macos.rb:14]: A test for this would be nice.

File: Library/Homebrew/tap.rb [hunk 4]
```diff
@@ -567,6 +570,12 @@ def audit_exceptions
     @audit_exceptions = read_formula_list_directory "#{HOMEBREW_TAP_AUDIT_EXCEPTIONS_DIR}/*"
   end
 
+  # Hash with style exceptions
+  sig { returns(Hash) }
+  def style_exceptions
+    @style_exceptions = read_formula_list_directory "#{HOMEBREW_TAP_STYLE_EXCEPTIONS_DIR}/*"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/tap.rb:576]: A test for this would be nice.

File: Library/Homebrew/tap.rb [hunk 5]
```diff
@@ -760,6 +769,15 @@ def audit_exceptions
     end
   end
 
+  # @private
+  def style_exceptions
+    @style_exceptions ||= begin
+      self.class.ensure_installed!
+      super
```
Comment (MikeMcQuaid) [inline Library/Homebrew/tap.rb:776]: A test for this would be nice.

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by Rylan12** [comment]
Alright, I added tests for the requested places. I'll take a look at the code coverage results and I'll add some more if it's too much work to do. Otherwise, I think this is good (looks like the PR has made it to linuxbrew-core, so assuming there are no surprises I should have this merged later tonight!

**Review by MikeMcQuaid — APPROVED** [review]

**Pull request merged**